language/vm/vm_runtime/src/interpreter.rs

```diff
@@ -0,0 +1,1204 @@
+                i += 1;
+            }
+            internal_state.push_str(format!("{}* {:?}\n", i, code[pc]).as_str());
+        }
+        internal_state
+            .push_str(format!("Locals:\n{}", current_frame.locals.pretty_string()).as_str());
+        internal_state.push_str("Operand Stack:\n");
+        for value in &self.operand_stack.0 {
+            internal_state.push_str(format!("{}\n", value.pretty_string()).as_str());
+        }
+        internal_state
+    }
+
+    /// Generate a core dump and an `UNREACHABLE` invariant violation.
+    fn unreachable(&self, msg: &str, current_frame: &Frame<'txn, FunctionRef<'txn>>) -> VMStatus {
+        let err = VMStatus::new(StatusCode::UNREACHABLE).with_message(msg.to_string());
+        self.maybe_core_dump(err, current_frame)
+    }
+}
+
+// TODO Determine stack size limits based on gas limit
+const OPERAND_STACK_SIZE_LIMIT: usize = 1024;
+const CALL_STACK_SIZE_LIMIT: usize = 1024;
+
+/// The operand stack.
+struct Stack(Vec<Value>);
+
+impl Stack {
+    /// Create a new empty operand stack.
+    fn new() -> Self {
+        Stack(vec![])
+    }
+
+    /// Push a `Value` on the stack if the max stack size has not been reached. Abort execution
+    /// otherwise.
+    fn push(&mut self, value: Value) -> VMResult<()> {
+        if self.0.len() < OPERAND_STACK_SIZE_LIMIT {
+            self.0.push(value);
+            Ok(())
+        } else {
+            Err(VMStatus::new(StatusCode::EXECUTION_STACK_OVERFLOW))
+        }
+    }
+
+    /// Pop a `Value` off the stack or abort execution if the stack is empty.
+    fn pop(&mut self) -> VMResult<Value> {
+        self.0
+            .pop()
+            .ok_or_else(|| VMStatus::new(StatusCode::EMPTY_VALUE_STACK))
+    }
+
+    /// Pop a `Value` of a given type off the stack. Abort if the value is not of the given
+    /// type or if the stack is empty.
+    fn pop_as<T>(&mut self) -> VMResult<T>
+    where
+        VMResult<T>: From<Value>,
+    {
+        self.pop()?.value_as()
+    }
+
+    /// Pop n values off the stack.
+    fn popn(&mut self, n: u16) -> VMResult<Vec<Value>> {
+        let remaining_stack_size = self
+            .0
+            .len()
+            .checked_sub(n as usize)
+            .ok_or_else(|| VMStatus::new(StatusCode::EMPTY_VALUE_STACK))?;
+        let args = self.0.split_off(remaining_stack_size);
+        Ok(args)
+    }
+}
+
+/// A call stack.
+struct CallStack<'txn>(Vec<Frame<'txn, FunctionRef<'txn>>>);
+
+impl<'txn> CallStack<'txn> {
+    /// Create a new empty call stack.
+    fn new() -> Self {
+        CallStack(vec![])
+    }
+
+    /// Return `true` if the call stack has room for a `Frame`, `false` otherwise.
+    fn can_push(&self) -> bool {
+        self.0.len() < CALL_STACK_SIZE_LIMIT
+    }
+
+    /// Push a `Frame` on the call stack.
+    // REVIEW: This call must be protected by a `can_push` call and we may want to assert something
+    fn push(&mut self, frame: Frame<'txn, FunctionRef<'txn>>) {
```

Comment:
The push/pop APIs of `CallStack` & `Stack` seem inconsistent to me. Is there a specific reason we have to manually protect it with `can_push`? Intuitively I would prefer `push` returning a result.

Reply:
yeah I hear you, that is the way I wrote it initially but it has issues, and this was the simplest way to do it for now.
The problem in mimicking the Stack API (so `push` that returns a `Result`) is that the `Frame` is moved in (so "consumed", passed not as a reference but "by value") and cannot be used in a core dump in the caller (given it's "gone"). So in order to keep the `Frame` alive in the caller we either have to use the model here or we need to return the `Frame` in the `Result` which would make the `Result` incompatible with a `VMResult` - and I like that solution even less.
This is all not ideal and it may change in the future a bit (e.g. the interpreter loop moving to `Frame`) so I figured it was good enough for now.
Thoughts?

Reply:
I see. Yeah I'm fine with the reason given.

Reply:
You can return the original frame back in the error type -- I can walk over how to do this in person